### PR TITLE
[3] #239 Sheet Bending

### PR DESCRIPTION
**Pull request by Erik9519**
**DO NOT MERGE UNTIL BASE BRANCH IS DEVELOPMENT**

#### What have I done
Sheet Model movement and beding

#### Testing
None

#### Addresses SRD section(s)
SheetModel

### CODE REVIEW

File: src/app/core/models/sheet-model.ts [hunk 1]
```diff
@@ -93,6 +94,48 @@ export class SheetModel {
...
+        // Bend Sheet
+        for (let i = 0; i < curvePoints.length; i++) {
+            // Compute Normal From Tangent
+            const normal = new THREE.Vector3(
+                curvePoints[i].getTangent().y,
+                -curvePoints[i].getTangent().x,
+                0
+            ).normalize().multiplyScalar(this.thickness / 2);
+            // Pre-Compute X and Y
+            const dir: THREE.Vector3 = new THREE.Vector3(
```
Comment (TijnBertens) [inline src/app/core/models/sheet-model.ts]: The name `dir` does not really accurately describe this variable, as it contains the model space coordinates of the curve point (relative to the leading edge)

File: src/app/core/models/sheet-model.ts [hunk 2]
```diff
@@ -93,6 +94,48 @@ export class SheetModel {
...
+                curvePoints[i].getPoint().x - curvePoints[0].getPoint().x,
+                curvePoints[i].getPoint().y - curvePoints[0].getPoint().y,
+                0);
+            // Displace Bottom Segment Vertex
+            this.geometry.vertices[(i * 2)].x = dir.x;
+            this.geometry.vertices[(i * 2)].y = dir.y;
+            this.geometry.vertices[(i * 2)].sub(normal);
+            // Copy to vertex on the other side
+            this.geometry.vertices[(i * 2) + 1].x = this.geometry.vertices[(i * 2)].x;
```
Comment (TijnBertens) [inline src/app/core/models/sheet-model.ts]: Perhaps move these calculations like `(i*2) + 1` into some const variable like `bottomVertexLeftIndex`, that way they only happen once and we can see what they mean.

File: src/app/core/models/sheet-model.ts [hunk 3]
```diff
@@ -93,6 +94,48 @@ export class SheetModel {
         this.model.position.set(0, 0, 0);
     }
 
+    /**
+     * Move and bend sheet along a set of points
+     * Sheet cannot be moved along z axis
+     * @param curvePoints the points onto which the sheet segments are to be snapped
+     */
+    moveTo(curvePoints: CurvePoint[]): void {
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Changed

File: src/app/core/models/sheet-model.ts [hunk 4]
```diff
@@ -93,6 +94,48 @@ export class SheetModel {
...
+        // Bend Sheet
+        for (let i = 0; i < curvePoints.length; i++) {
+            // Compute Normal From Tangent
+            const normal = new THREE.Vector3(
```
[the remark this answers is not in the document]
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Changed

File: src/app/core/models/sheet-model.ts (hunk 1 above)
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Now called position

File: src/app/core/models/sheet-model.ts (hunk 2 above)
Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Changed

File: src/app/core/models/sheet-model.ts [hunk 5]
```diff
@@ -93,6 +95,59 @@ export class SheetModel {
         this.model.position.set(0, 0, 0);
     }
 
+    /**
+     * Move and bend sheet along a set of points
```
Comment (TijnBertens) [inline src/app/core/models/sheet-model.ts]: Outdated

Comment (Erik9519) [reply src/app/core/models/sheet-model.ts]: Fixed

**Pull request merged**